### PR TITLE
[None][perf] fuse Kimi K3 routing and MXFP8 quantization

### PR DESCRIPTION
**Pull request by longlee0622**
## What changed

- Add a Kimi K3 decode-specialized CUDA kernel that runs 896-expert/top-16 no-aux routing and BF16-to-MXFP8 quantization in one launch for token counts up to 64.
- Reuse the existing DeepSeek V3 routing device implementation and the existing MXFP8 conversion helper.
- Wire the fast path into the external-communication scheduler for the TRTLLMGen W4A8 MXFP4/MXFP8 backend.
- Keep all unsupported shapes, routing modes, communication modes, EPLB configurations, and backends on the existing unfused path.
- Enable the kernel on SM10x architectures (`100 <= SM < 110`). Set `TLLM_K3_DISABLE_FUSED_ROUTE_QUANT=1` to restore the unfused path.

## Why

Kimi K3 decode currently launches no-aux routing, converts the routing scales, and launches MXFP8 activation quantization separately before the MoE op. At small decode batches, launch overhead is significant and is repeated across every routed MoE layer. Routing and activation quantization are independent per-token work, so they can share one grid and execute concurrently.

This optimization is specific to the TRTLLMGen backend. `MEGAMOE_DEEPGEMM` uses the fused-communication scheduler and its own `megamoe_prepare` path, so it does not call this fast path.

## Validation

- B200 parity against the existing `noaux_tc_op + mxfp8_quantize` chain for `M={1,5,64}`: exact expert indices, BF16 routing scales, FP8 bytes, and UE8M0 scales.
- Kimi gate test file: 11 tests passed in the B200 development container.
- B200 microbenchmark at `M=5`: approximately 35.8 us unfused versus 22.4 us fused.
- Repository pre-commit hooks passed on the final commits.

The SM103/SM107 runtime gate is source-compatible but has not yet been validated on those GPUs in this PR.

## Post-#17088 integration

- Rebased manually onto the new `ExternalCommMoEScheduler` and `do_finalize` interfaces.
- Preserved the external-communication TRTLLMGen fast path.
- If #17088 supplies an `MxFp8QuantizedTensor`, the route+quant helper returns before the SM probe and does not quantize again.
- Added a CPU regression for that pre-quantized handoff.
- Final head `181fb03e15`: local and GitHub pre-commit checks plus DCO pass.

The B200 parity/microbenchmark data above predates this scheduler rebase. Post-rebase external-communication end-to-end validation remains recommended.

## Dev Engineer Review

- Added a Kimi K3 fused route-and-MXFP8-quantization CUDA kernel for 896 experts, top-16 routing, and up to 64 tokens.
- Added CUDA and Torch APIs with SM10x, device, dtype, contiguity, shape, and token-count validation.
- Integrated the fused path into the TRTLLMGen external-communication scheduler.
- Preserved the unfused path for unsupported configurations.
- Added the `TLLM_K3_DISABLE_FUSED_ROUTE_QUANT=1` opt-out.
- Added `MxFp8QuantizedTensor` support for MXFP8 data, scales, and row splitting.
- Added fake-operator support for tracing and compilation.
- B200 parity validation passed. Benchmark results improved from approximately 35.8 µs unfused to 22.4 µs fused at `M=5`.
- Review focus should include CUDA stream serialization, output dtype and layout consistency, MXFP8 scale encoding, scheduler dispatch ordering, and partial fused-output handling.
- SM103/SM107 runtime validation and external-communication end-to-end validation remain outstanding.

## QA Engineer Review

- Added `test_fused_route_quant_matches_unfused_chain(num_tokens)` in `tests/unittest/_torch/modules/moe/test_kimi_k3_moe_gate.py`.
  - Covers token counts 1, 5, and 64.
  - Listed in `tests/integration/test_lists/test-db/l0_b200.yml`.
- Added `test_kimi_fused_route_quant_skips_prequantized_input(monkeypatch)` and `test_kimi_mxfp8_quantized_tensor_handoff()` in `tests/unittest/_torch/modules/moe/test_moe_backend.py`.
  - These tests are not listed in a `test-db/` or `qa/` file.
- The parity test covers routing outputs, routing scales, quantized hidden states, and quantization scales.
- The pre-quantized handoff tests cover CPU-side backend selection and MXFP8 payload preservation.
- Verdict: needs follow-up. Add CI or manual-QA coverage for the pre-quantized handoff tests and complete SM103/SM107 and external-communication runtime validation.

### CODE REVIEW

**Comment by xxi-nv** [comment]
@sunnyqgg Please help to reiview this PR.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py [hunk 1]
```diff
@@ -583,6 +583,43 @@ def load_weights(self,
...
+        router_logits: torch.Tensor,
+    ) -> Optional[Tuple[torch.Tensor, torch.Tensor, torch.Tensor,
+                        torch.Tensor]]:
+        """Fuse Kimi K3 no-aux routing and MXFP8 input quantization.
+
+        This launch-overhead optimization is deliberately specialized to the
+        K3 decode shape. Returning ``None`` keeps every other model, shape,
+        architecture, and op backend on the existing unfused path.
+        """
+        sm_version = get_sm_version()
```
Comment (sunnyqgg) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py]: get_sm_version() runs before the MxFp8QuantizedTensor / env-flag checks, so test_kimi_fused_route_quant_skips_prequantized_input (which patches it to raise) fails deterministically. Please move the cheap checks before the SM probe

File: tensorrt_llm/_torch/modules/fused_moe/moe_scheduler.py [hunk 2]
```diff
@@ -397,11 +397,31 @@ def _forward_chunk_impl(
             or moe.comm is not None
             or FORCE_SEPARATED_ROUTING
         )
+        used_fused_route_quant = False
         if requires_separated_routing:
-            # Separated routing: ConfigurableMoE calls routing_method
-            token_selected_experts, token_final_scales = moe.routing_method.apply(
-                router_logits, input_ids
+            can_quantize_before_dispatch = (
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/fused_moe/moe_scheduler.py]: can_quantize_before_dispatch (l.402) and supports_post_quant (l.545) are the same predicate computed twice; consequently the assert not used_fused_route_quant at l.560 is unreachable — it guards drift between the two, not any runtime state. Suggest hoisting the predicate into one variable used by both sites and dropping the assert.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py [hunk 3]
```diff
@@ -29,7 +29,7 @@
     fp4_block_scale_fake_output_without_finalize
 from ...model_config import ModelConfig
 from ...utils import (ActivationType, ActType_TrtllmGen, AuxStreamType,
-                      Fp4QuantizedTensor)
+                      Fp4QuantizedTensor, MxFp8QuantizedTensor)
```
Comment (BowenFu) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py:32]: Where is `MxFp8QuantizedTensor` defined? The exact-head `tensorrt_llm._torch.utils` exports only `Fp4QuantizedTensor`, so this import fails before any MoE test can run. Please add the wrapper and handle its payload fields separately, or remove this incomplete handoff integration.

Comment (longlee0622) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py:32]: Good catch. This PR used to target feat/kimi_k3 and I retargeted it to main since the k3 -> main merge back is almost done. It seems the `MxFp8QuantizedTensor` is added to k3 feature branch in this PR but never merged back to main: https://github.com/NVIDIA/TensorRT-LLM/pull/17088/changes#diff-f2f9a3f5a11b927dc699020f15b174d43cb7fa6d3497a56dca733b8e7116960aR184

@brnguyen2 Do you have any other merge-back PRs covering this part?

File: tensorrt_llm/_torch/utils.py [hunk 4]
```diff
@@ -195,6 +195,37 @@ def shape(self):
         return self.fp4_tensor.shape
 
 
+@dataclass
+class MxFp8QuantizedTensor:
+    """MXFP8 activation and per-1x32 UE8M0 scaling factors."""
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/_torch/utils.py]: +1

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py (hunk 3 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_trtllm_gen.py:32]: No, I don't have any other merge-back PRs covering this at the moment (tracked under TRTLLM-14816). But it looks like this PR was updated to add `MxFp8QuantizedTensor` directly in `utils.py`, so it can land standalone on main without depending on anything else.

Heads up though: TRTLLM-14816 ports the same perf pack from the feature branch to main, which includes this fused route+quant work, so we'll want to coordinate so the class and kernel don't land twice.

**Review by BowenFu — APPROVED** [review]

File: tensorrt_llm/_torch/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (longlee0622) [reply tensorrt_llm/_torch/utils.py]: Fixed. MxFp8QuantizedTensor now has a dedicated handoff path using fp8_tensor, while the existing FP4 path continues to use fp4_tensor. I also added the requested interface annotations/layout documentation and a focused regression test, and wired both pre-quantized handoff tests into l0_b200.yml.

**Review by BowenFu — APPROVED** [review]

**Review by kris1025 — APPROVED** [review]
